Replace webpack & rollup with vite for avivator & viv builds

manzt opened the pull request:
I just updated another project using webpack with Vite, and I thought I'd give it a try with Viv.

This PR strips out _all_ the dev dependencies for Avivator in favor of using [vite](https://vitejs.dev/) for the complete build. Vite is a development server and production bundler, but all of development is "un-bundled". The vite dev-server tracks changes in both `src/` and `avivator/src`, so there is no longer the need to run `rollup` concurrently with the `webpack-dev-server`.

**TL;DR**

```bash
npm start # `vite` === starts avivator dev server
npm run build # `vite build --mode lib` === bundles the viv library source code into `dist`
npm run build:avivator # `vite build` === bundles a production build of avivator in `avivator/dist`
```

The biggest required change is that any files containing JSX must use `.jsx` rather than `.js` extensions. Second, vite now inlines the [`Web Workers`](https://vitejs.dev/guide/features.html#web-workers) rather than the rollup-plugin. However, inlining only happens during production and in development worker modules are used which are sadly only implemented in [chromium browsers](https://developer.mozilla.org/en-US/docs/Web/API/Worker/Worker). 

If you are interested @ilan-gold , feel free to checkout and try it out. Otherwise, I can close!

manzt commented on package.json:
```diff
@@ -1,23 +1,22 @@
   "version": "0.10.4",
-  "main": "dist/index.js",
-  "module": "dist/index.js",
+  "main": "dist/viv.es.js",
+  "module": "dist/viv.es.js",
```
I can't figure out how to change the name of the final build from vite. We can always run `mv dist/viv.es.js dist/index.js` if desired.

manzt commented:
The eslint warnings are different from on my machine? have you noticed this before @ilan-gold

ilan-gold commented:
Checking it out now - hopefully even if we don't end up using this we can translate some lessons to Vitessce as @keller-mark has tried Vite in the past for us there and had some success (although it's a mess right now so there's work left to be done there).  Vitessce badly needs the overhaul.

keller-mark commented:
In Vitessce I did get the development server working with Vite at one point https://github.com/vitessce/vitessce/compare/keller-mark/vite but the hot refresh was extremely slow to the point where it was unusable. Maybe it has been improved by now though, or maybe I was doing something wrong.  I currently don't have the bandwidth to look at it again but maybe in a few months

ilan-gold commented:
Thanks @keller-mark - didn't mean that your work was a mess, more that the current state of things is a bit of a mess having tried to replace the bundling myself.

ilan-gold commented on rollup.config.test.js:
```diff
@@ -0,0 +1,21 @@
+import resolve from '@rollup/plugin-node-resolve';
```
Why do we need to use rollup for the tests?

manzt replied:
Our source code uses node-isms for dependency structure (not pure ESM): 

```
submodule/
    |-- index.js
```

and importing like:

```javascript
import { a, b } from './submodule';
```

to remove this dependency, we would need to change these imports to:

```javascript
import { a, b } from './submodule/index';
```

manzt replied:
Oh sorry, I thought this was an issue with `@rollup/plugin-node-resolve` not `rollup` at all.

Our previous testing approach required bundling our source code for tests. This is still required because 1.) our source isn't actually ESM (so we can't run node directly), and 2.) we need to pipe CJS in browserify for `tape-run` for the electron tests.

manzt replied:
FWIW vite uses rollup & rollup-plugins under the hood, it's just I can't pipe the output to std out like we have previously.

ilan-gold reviewed (approved):
>  Second, vite now inlines the [`Web Workers`](https://vitejs.dev/guide/features.html#web-workers) rather than the rollup-plugin. However, inlining only happens during production and in development worker modules are used which are sadly only implemented in [chromium browsers](https://developer.mozilla.org/en-US/docs/Web/API/Worker/Worker).

Can you explain this a bit more?  You're saying web workers don't work at all outside chromium?  They seem to be working in FF.  I would be hesitant to do this if something were broken, but nothing appears to be broken despite your comment.

Also, could you test the build out in an upstream application like Vitessce and/or Vizarr to be sure nothing has broken that we are missing?

Other than that, I don't see a reason not to do this.  Things seem much simpler and our linting appears to be working again, at least on Github CI.  Thanks!

ilan-gold commented:
Just realized you never formally requested review.  Feel free to request again if needed.

manzt commented:
The _developement_ server approach for both `vite` _and_ `snowpack` is completely "un-bundled". This means that modules in the source code are all treated as separate modules; there is _zero_ bundling of your project source code that occurs prior to "build". This means that during development files are transformed individually, and any `import` in your source code ends up as an import or request that your browser makes to the dev server.

The issue is that using import statements inside of Web Workers is not supported across all browsers. E.g.

```javascript
// worker.ts
import { getDecoder } from '/geotiff/compression/index.js';
// setup web-worker
```

only works in chromium browsers, or browsers that support initializing web-workers like:

```javascript
const worker = new WebWorker('./worker.js', { type: 'module' });
```

There is an outstanding issue for this in FF: https://bugzilla.mozilla.org/show_bug.cgi?id=1558780

To be clear totally clear, this is only an "issue" when running the development server. Running `vite build` will create a production build of the website in `avivator/dist` where the imports are bundled _and_ inlined for the workers. You can preview this build in _any_ browser using `vite preview`.

ilan-gold commented:
I guess my confusion is that nothing appears to be broken to the end of user of `npm start` but what you're saying makes it seem like something should be (more noticeably) broken.  If the only difference is that workers are not in-lined in FF during development, that's fine.  But it sounds like something different?  I just want to understand this, sorry for my being slow!

manzt commented:
<img width="839" alt="image" src="https://user-images.githubusercontent.com/24403730/125104547-80ced480-e0ab-11eb-920b-b8bab8d2db26.png">

ilan-gold commented:
Ok I see that.  So how does the application work in FF then?

ilan-gold commented:
> Not quite. The worker(s) will not work in dev mode in FF, but it seems that somehow the data are decoded regardless (no decoders).

Ok, so the decoding is not parallelized?

ilan-gold commented:
I see the pool being instantiated with `Worker` objects
<img width="648" alt="Screen Shot 2021-07-09 at 5 54 26 PM" src="https://user-images.githubusercontent.com/43999641/125105527-d670a480-e0de-11eb-93f9-4e6eda017ad1.png">

manzt commented:
Again, I'm not quite sure what is going on. The error is thrown from _inside_ the worker, so the Pool object will look normal. It's just the code inside throws.

ilan-gold commented:
Perhaps unrelatedly, I am not sure the proxy is working for the pool, or perhaps I am missing something:

```javascript
export function createPoolProxy(tiff: GeoTIFF, pool: Pool) {
  const get = (target: GeoTIFF, key: any) => {
    // Intercept calls to `image.readRasters`
    if (key === 'readRasters') {
      return (options: Parameters<typeof target.readRasters>) => {
        console.log(pool)
        // Inject `pool` argument with other raster options.
        return target.readRasters({ ...options, pool });
      };
    }

    // tiff['__viv-decoder-pool'] === true
    if (key === POOL_PROXY_KEY) {
      return true;
    }

    return Reflect.get(target, key);
  };
  return new Proxy(tiff, { get });
}
```

The `console.log` statement never seems to fire.  It does seem to work with the offsets proxy and when I add a `console.log(pool)` to the geotiff source code, it returns `null`.

manzt commented:
I fixed the decoder in #470 

I added an extra "dev-only" plugin for vite to bundle the worker with `esbuild`. It removes the errors from the console.

ilan-gold reviewed (commented):
Can you update the Avivator README.md? Thanks!  Looks good!
